src/typing/type_inference_js.ml

```diff
@@ -1292,8 +1292,29 @@ and toplevels cx stmts =
       if uc < List.length stmts
       then (
         let msg = "unreachable code" in
-        let (loc, _) = List.nth stmts uc in
-        Flow_js.add_warning cx [mk_reason "" loc, msg]
+        let rec drop n lst =
+          match (n, lst) with
+          | (_, []) -> []
+          | (0, l) -> l
+          | (x, h :: t) -> drop (x-1) t
+        in
+        let trailing = drop uc stmts in
+        trailing |> List.iter Ast.Statement.(fun stmt ->
+          match stmt with
+          (* function declarations are hoisted, so not unreachable *)
+          | (_, FunctionDeclaration {
+                FunctionDeclaration.id;
+                params; defaults; rest;
+                body;
+                returnType;
+                typeParameters;
+                async;
+                _
+              } ) -> ()
+          (* TODO also skip variable declarations if they do not have
+                  any assignments *)
+          | (loc, _) -> Flow_js.add_warning cx [mk_reason "" loc, msg]
```

Comment:
I'm pretty new to OCaml myself, but here's what I came up with:
```
| (_, VariableDeclaration { VariableDeclaration.declarations })
  when declarations |> List.for_all (function
    | (_, { VariableDeclaration.Declarator.init; _ }) -> Option.is_none init
  ) -> ()
```

However, I think it would be even nicer if we pointed to the `loc` of the init itself, so maybe something more like this:

```
(* variable declarations are hoisted, but initializers are not *)
| (_, VariableDeclaration { VariableDeclaration.declarations }) ->
    declarations |> List.iter VariableDeclaration.Declarator.(function
      | (_, { init = Some (init_loc, _); _ }) ->
        Flow_js.add_warning cx [mk_reason "" init_loc, msg]
      | _ -> ()
    )
```